lib/spack/spack/cmd/__init__.py

```diff
@@ -226,13 +274,98 @@ def disambiguate_spec_from_hashes(spec, hashes, local=False,
             database query. See ``spack.database.Database._query`` for details.
     """
     if local:
-        matching_specs = spack.store.db.query_local(spec, hashes=hashes,
-                                                    installed=installed)
+        pre_matching_specs = spack.store.db.query_local(spec, hashes=hashes,
+                                                        installed=installed)
     else:
-        matching_specs = spack.store.db.query(spec, hashes=hashes,
-                                              installed=installed)
+        pre_matching_specs = spack.store.db.query(spec, hashes=hashes,
+                                                  installed=installed)
+
+    matching_specs = []
+    conflicting_specs = []
+    narrowed_specs = []
+    loaded_specs = _find_already_loaded()
+
+    if compatible:
+        # compute the lists of conflicting and matching specs
+        loaded_hashes = [loaded_specs[name].dag_hash()
+                         for name in loaded_specs]
+        for possible_spec in pre_matching_specs:
+            conflict = False
+            for node in possible_spec.traverse():
+                if (loaded_specs.get(node.name) and
+                    (loaded_specs[node.name].dag_hash() !=
+                     node.dag_hash())):
+                    possible_spec._message = node.name
+                    conflict = True
+                    break
+            if possible_spec.dag_hash() in loaded_hashes:
+                tty.die(str(possible_spec.name) + ' is alreaded loaded\n')
+            if conflict:
+                conflicting_specs.append(possible_spec)
+            else:
+                matching_specs.append(possible_spec)
+    else:
+        matching_specs = pre_matching_specs
+
+    if preferred:
+        # see what the preferred spec would be,
+        # compare against the previously matched versions
+        # if none of them match, allow the whole list
+        # through, as we want to honor compatible specs and/or
+        # what the user specified over the preferences
+        if len(matching_specs) > 1:
+            prefspec = _apply_defaults(spec)
+            for possible_spec in matching_specs:
+                if possible_spec.satisfies(prefspec):
+                    narrowed_specs.append(possible_spec)
+            if len(narrowed_specs) != 0:
+                matching_specs = narrowed_specs
+
     if not matching_specs:
-        tty.die("Spec '%s' matches no installed packages." % spec)
```

Comment:
I'm not as confident as I'd like to be off the top of my head, but I *think* you could replace this whole block with a call to `spack.concretize.concretize_specs_together`, which you can see is used in `lib/spack/spack/environment.py` in the method `_concretize_together`.

I think this algorithm cannot be guaranteed to work. In your code, you may produce an incompatible spec. If you use `concretize_together`, you'll get an error and you'll need to decide what to output in case of error. (I suggest requesting the user turn off the compatibility check feature, while presenting the option of unloading some of the loaded packages. I don't think the concretizer can currently tell them which ones to unload, but I could be wrong about that.)